lib/core-server/src/utils/manager-cache.ts

```diff
@@ -1,22 +1,59 @@
+import { Options } from '@storybook/core-common';
+import { logger } from '@storybook/node-logger';
+import fs from 'fs-extra';
+import path from 'path';
 import { stringify } from 'telejson';
 import webpack from 'webpack';
-import { FileSystemCache } from 'file-system-cache';
+
+// The main config file determines the managerConfig value, so is already handled.
+// The other files don't affect the manager, so can be safely ignored.
+const ignoredConfigFiles = [/^main\.(m?js|ts)$/, /^preview\.(m?js|ts)$/, /^preview-head\.html$/];
 
 export const useManagerCache = async (
-  fsc: FileSystemCache,
+  cacheKey: string,
+  options: Options,
   managerConfig: webpack.Configuration
 ) => {
+  const [cachedISOTime, cachedConfig] = await options.cache
+    .get(cacheKey)
+    .then((str) => str.match(/^([0-9TZ.:+-]+)_(.*)/).slice(1))
+    .catch(() => []);
+
   // Drop the `cache` property because it'll change as a result of writing to the cache.
   const { cache: _, ...baseConfig } = managerConfig;
   const configString = stringify(baseConfig);
-  const cachedConfig = await fsc.get('managerConfig').catch(() => {});
-  await fsc.set('managerConfig', configString);
-  return configString === cachedConfig;
+  await options.cache.set(cacheKey, `${new Date().toISOString()}_${configString}`);
+  if (configString !== cachedConfig || !cachedISOTime) {
+    logger.line(1); // force starting new line
+    logger.info('=> Ignoring cached manager due to change in manager config');
+    return false;
+  }
+
+  // Check the modification time for all files in the config dir (.storybook) and
+  // don't use the cache if any file has been modified since the cache was created.
+  const configFiles = await fs.readdir(options.configDir);
+  const cacheCreationDate = new Date(cachedISOTime);
+  try {
+    await Promise.all(
+      configFiles.map(async (file) => {
+        if (ignoredConfigFiles.some((pattern) => pattern.test(file))) return;
+        const filepath = path.join(options.configDir, file);
+        const { mtime: fileModificationDate } = await fs.stat(filepath);
+        if (fileModificationDate > cacheCreationDate) throw filepath;
```

Comment:
How do we guarantee that this is `false` on the very first execution?

Reply:
At first run, there won't be a `cachedISOTime`, so the function will return before reaching this point. See https://github.com/storybookjs/storybook/pull/14993/files#diff-fbd9d4dafd5b84732adcde34aa949c01bcb44e293b5adf8cb512c0bfdcc1cc45R26